## Genomics England HTSget example for GLH usage

**WARNING!** Direct use of the GeL HTSget service is currently not part of NGIS and subject to potential change without notice

In the following examples the "UAT" instance of HTSget has been used

| Env  | OPENCGA HSCN URL                                                     | HTSget HSCN URL                                 |
| ---- | ---------------------------------------------------------------------| ----------------------------------------------- |
| test | https://opencga-gms-test.genomicsengland.nhs.uk/opencga/webservices/ | https://htsget-gms-test.genomicsengland.nhs.uk/ |
| uat  | https://opencga-gms-beta.genomicsengland.nhs.uk/opencga/webservices/ | https://htsget-gms-beta.genomicsengland.nhs.uk/ |
| prod | https://opencga.genomicsengland.nhs.uk/opencga/webservices/          | https://htsget.genomicsengland.nhs.uk/          |

Before using HTSget you need to get an access token from OpenCGA, save it, and set its location as an environment variable

To do this you will need a username and password from Genomics England with permissions to access files in OpenCGA

## Get Access Token from OpenCGA

This example is for uat environment

```
curl -X POST --header "Content-Type: application/json" --header "Accept: application/json" --header "Authorization: Bearer " -d "{\"password\": \"$PASSWORD\"}" "https://opencga-gms-beta.genomicsengland.nhs.uk/opencga/webservices/rest/v1/users/$USERNAME/login" | jq '.response[0].result[0].token' | sed -e 's/"//g' > ~/.hts_token
```

and then, set the location of this token as an environment variable which samtools/bcftools will refer to:

```
export HTS_AUTH_LOCATION=~/.hts_token
```

**Note!** This token will only be valid for ~30 mins and you will need to rerun the curl command to refresh it after then (but you only need to run the export command once)

**Note!** The following commands require you to have samtools/bcftools installed locally

### Example streaming a bam / cram using HTSget "reads" endpoint (uat environment) for a specific chrom, start, end region
```
samtools view 'https://htsget-gms-beta.genomicsengland.nhs.uk/reads/bypath/1053593329/by_name:LP5100038-DNA_C09:BE00005345:LP5100038-DNA_C09.cram?referenceName=22&start=23180509&end=23280509' > example.sam
```
### Example streaming a VCF using HTSget "variants" endpoint (uat environment) for a specific chrom, start, end region
```
bcftools view 'https://htsget-gms-beta.genomicsengland.nhs.uk/variants/bypath/1053593329/analysis:multisample:e07f18889c47b16ab654c67512765481:944cb0f0bc3fa81bb06784106bd0be30:split_joint_vcf:3d7077fd923993aefa9044a384764b0d:1:output:r23473011521_10000_LP5100038-DNA_C09.vcf.gz?referenceName=22&start=23180509&end=23280509' > example.vcf
```
### Obtaining file paths

CIP-API interpretation request payloads contain the details of the files and their associated file paths that are generated by the pipeline and dispatched to DSS applications.

Therefore, HTSget file paths for bams / crams & vcfs can be obtained from CIP-API payloads - see `get_file_paths_for_htsget.py` for examples

Please refer to the online CIPAPI user guide https://ip-cipapi-help.genomicsengland.co.uk/3.6/usage/#interpretation_request_data for further information about obtaining CIP-API payloads
